### PR TITLE
Expand the README

### PR DESCRIPTION
**Pull request by lerks**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -72,6 +135,23 @@ Backends come in two flavors:
   saturate the NIC's bandwidth. Or using a CUDA memcpy call to transfer memory
   from one GPU to another using NVLink.
 
+These different usage patterns promote different design choices when
+implementing transports and channels, which means the two are not perfectly
+interchangeable. For example, a TCP-based transport is best implemented using a
```
Comment (jiayisuse) [inline README.md]: Shm-based? :)

Comment (lerks) [reply README.md]: No, I did mean TCP, it's a discussion about the tradeoffs of using one vs more than one sockets for the two use cases.

File: README.md [hunk 2]
```diff
@@ -41,6 +41,69 @@ There are four classes you need to know about:
...
+modifying it. In other terms, you relinquish control over the memory when you
+pass a message to the pipe, only to reacquire it once the message is given back
+to you in the callback. This contract is encoded by the requirement to move the
+messages into and out of the pipe (using rvalue references). Also, because of
+this agreement, all callbacks will always be called, even if the pipe is closed
+or if it errors, in order to give back the memory.
+
+The order in which messages are written to a pipe is preserved when these
+messages are read on the other side. Moreover, for a given pipe endpoint, the
+callbacks of the performed operations are executed in the same order that these
```
Comment (javier-m) [inline README.md]: what are write callbacks for?
maybe it would be easier to make two sections in the README: 1) on `write` 2) on `read`

Comment (javier-m) [reply README.md]: as well as some written examples to illustrate how to use the API maybe (e.g., at the beginning of the README, tl;dr style)

File: README.md [hunk 3]
```diff
@@ -41,6 +41,69 @@ There are four classes you need to know about:
...
   tensors (memory living on any device, like GPUs).
 
+Sending a message from one end of the pipe to the other can be achieved using
+the `write` method, which takes a message (with the data to send) and a
+callback which will be invoked once the sending has completed. This callback
+will be invoked with an error (if one happened) and with the message.
+
+Receiving a message takes two steps. First, you must register a callback that
+will be notified for incoming messages. This is performed by calling the
+`readDescriptor` method with said callback. The callback will be invoked with
```
[the remark this answers is not in the document]
Comment (lerks) [reply README.md]: Well, yes, the client cannot possibly provide a callback that is run by the server and has access to the server's memory. Anyway, I added a sentence about the fact that the descriptor contains metadata that can be used for allocation hints.

**Pull request closed without merging**